src/components/widgets/accounts-balance-sheet/accounts-balance-sheet.directive.coffee

```diff
@@ -100,20 +99,27 @@ module.controller('WidgetAccountsBalanceSheetCtrl', ($scope, $q, ImpacWidgetsSvc
         return true
 
   translateCategories = (categories) ->
+
     _.each(categories, (category) ->
+      # $scope.categories needs to be filled with keys before sortData() runs.
+        # Pushing the categories after $translate resolves would be too late.
+      newCategory = {key: category}
+      $scope.categories.push(newCategory)
```

Comment:
I think we can avoid having future developers proceed with caution around this, by maybe having a look at how the component is initialised, and handling the asynchronicity of this method with promises.

When you say "$scope.categories needs to be filled with keys before sortData() runs", is that on load? If so, do you think it would be worth doing something like below:

```coffeescript
  w.initContext = ->
    if $scope.isDataFound = angular.isDefined(w.content) && !_.isEmpty(w.content.summary) && !_.isEmpty(w.content.dates)
      ...
      # While we are here I think we should remove this for the pattern below instead
      # $scope.categories = []
      # translateCategories should use _.map not _.each to return an Array of translated categories
      translateCategories(Object.keys(w.content.summary)).then((categories)->
          $scope.categories = categories
      ).finally(->
        initDates()
        sortData()
      )
    else
      initDates()
      sortData() 
```

I can see `$scope.sort` calls the sortData function, so might have to add some guarding for that as to only run once w.initContext has finished like so:

```coffeescript
sortData = -> 
  return if _.isEmpty($scope.categories)
  ...
```

Reply:
I'll have a look -- but do agree that handling this with promises is more semantic.